engine/classes/ElggUser.php

```diff
@@ -47,36 +45,31 @@ public function getLanguage($fallback = null) {
 	 * {@inheritdoc}
 	 */
 	public function __set($name, $value) {
-		if (!array_key_exists($name, $this->attributes)) {
-			parent::__set($name, $value);
-			return;
-		}
-
 		switch ($name) {
-			case 'prev_last_action':
-			case 'last_login':
-			case 'prev_last_login':
-				if ($value !== null) {
-					$this->attributes[$name] = (int) $value;
-				} else {
-					$this->attributes[$name] = null;
-				}
-				break;
-
 			case 'salt':
 			case 'password':
-				_elgg_services()->logger->error("User entities no longer contain salt/password");
-				break;
-
-			// setting this not supported
+				_elgg_services()->logger->error("User entities no longer contain {$name}");
+				return;
 			case 'password_hash':
 				_elgg_services()->logger->error("password_hash is a readonly attribute.");
+				return;
+			case 'email':
+				if (!validate_email_address($value)) {
+					throw new \InvalidParameterException("Email is not a valid email address");
+				}
 				break;
-
-			default:
-				parent::__set($name, $value);
+			case 'username':
+				if (!validate_username($value)) {
+					throw new \InvalidParameterException("Username is not a valid username");
+				}
+				$existing_user = get_user_by_username($value);
```

Comment:
Shouldn't this check disabled entities as well?

Reply:
good point... need to fix that

Reply:
Let's add public API to check username availability. I keep reinventing the wheel every time